### ISSUE TITLE
when determining whether a unit file is enabled, ignore .wants/ symlinks pointing to invalid unit file paths, such as /dev/null

### ISSUE THREAD

**Issue by twojstaryzdomu** [systemd/systemd#23341]
**systemd version the issue has been seen with**
 > 250.4 (Debian), 246 (openSUSE)

**Used distribution**
 > Debian, openSUSE

**Linux kernel version used** (`uname -a`)
 > 5.15.38 (Debian) & 5.3.18 (openSUSE), issue seems to be independent of kernel release

**CPU architecture issue was seen on**
 > armv6l & x86_64, issue seems to be platform-agnostic

**Expected behaviour you didn't see**
 > systemctl status examining the link points to a file that does not contain any service definition

**Unexpected behaviour you saw**
 > systemd reports misleading service enabled with symlink pointing to /dev/null

**Steps to reproduce the problem**
Create/remove a symlink pointing to /dev/null under /etc/systemd/system/multi-user.target.wants to black out an existing service. Observe its impact on systemctl status. Reload-daemon after removing/creating the symlink. The blacked-out service does not start as intended, this seems to be a service status reporting problem only.

**Additional program output to the terminal or log subsystem illustrating the issue**

```
# rm -rf /{etc,lib}/systemd/system/multi-user.target.wants/test.service /lib/systemd/system/test.service 
# systemctl status test
Unit test.service could not be found.
# cat > /lib/systemd/system/test.service << EOL
[Service]
ExecStart=true
[Install]
WantedBy=multi-user.target
EOL
# systemctl daemon-reload
# systemctl status test
○ test.service
     Loaded: loaded (/lib/systemd/system/test.service; disabled; vendor preset: enabled)
     Active: inactive (dead)
# ln -sf /dev/null /etc/systemd/system/multi-user.target.wants/test.service
# systemctl daemon-reload
# systemctl status test
○ test.service
     Loaded: loaded (/lib/systemd/system/test.service; enabled; vendor preset: enabled)
     Active: inactive (dead)
```

**Comment by poettering** [comment]
the .wants/ symlinks are about adding deps, they do not affect the definition of the unit as a whole. Hence this works as intended. If you want to mask a unit, mask the unit file itself, i.e. in /etc/systemd/system/ instead of a .wants/ subdir below it.

Closing, since this works as intended really.

**Comment by twojstaryzdomu** [comment]
> the .wants/ symlinks are about adding deps, they do not affect the definition of the unit as a whole. Hence this works as intended.

I beg to differ, so you propose to break ("mask the unit file itself") a service altogether. Isn't that throwing the baby out with the bath towel?

What if you need to use the service, just not under that specific `target/.wants` directory?

The use case is blackmasking a unit under one target and no other. There are plenty of distro provisioned dependencies under (`/lib/systemd/...`) that you may want to disable selectively per target basis.

`/dev/null` links are quite common to manage such instances. I'd expect systemd not to present misleading status.

Hence, this bug is still present.

**Comment by mbiebl** [comment]
Right. While it's not clear what your use case is (trying to add /dev/null symlink in a wants directory appears strange), systemctl should not report misleading information, i.e. showing it as enabled.

**Comment by poettering** [comment]
> What if you need to use the service, just not under that specific `target/.wants` directory?

i cannot parse this. What is this supposed to mean?

You want to cancel individual deps? We have no mechanism for that. And its hard to do, since deps can be created in both directions and via various means.

**Comment by mbiebl** [comment]
Just to better understand your use case better:
You want a service that has been enabled satically via `/lib/systemd/system/multi-user.targets.wants/foo.service` disable via a `/etc/systemd/system/multi-user.targets.wants/foo.service`, to prevent it from being started "during boot".
But you don't want to mask it completely to be able to start the service manually?

**Comment by poettering** [comment]
#4830 is the issue where we track the request to be able to mask indivudual deps. If this is what is requested here, we can close this one.

**Comment by twojstaryzdomu** [comment]
> Just to better understand your use case better:
> You want a service that has been enabled statically via `/lib/systemd/system/multi-user.targets.wants/foo.service` disable via a `/etc/systemd/system/multi-user.targets.wants/foo.service`, to prevent it from being started "during boot"?
> But you don't want to mask it completely to be able to start the service manually?
Precisely. Suppose `/lib/systemd` is read-only and the only means to interact with its contents is via blackmasking under `/etc/systemd`. As systemd is building its fragments first from the contents of `/lib/systemd`, then the contents of `/etc/systemd`, systemd upon finding a service link pointing at `/dev/null` considers a service disabled under that specific target. It's as simple as that.

**Comment by mbiebl** [comment]
I think there are two issues here: 
- The misleading status output. systemctl shows the service as enabled when actually it isn't
Not sure what the current status of such a `/dev/null` symlink would mean, a misconfiguration probably. But certainly not `enabled`. `disabled` would probably be a better representation.
- and #4830

**Comment by twojstaryzdomu** [comment]
> * Not sure what the current status of such a `/dev/null` symlink would mean, a misconfiguration probably. But certainly not `enabled`. `disabled` would probably be a better representation.

Indeed, considering the presence of a link as a sure fire sign a service is at all enabled, without looking at its contents, or where it points to, is dodgy at best.

**Comment by mbiebl** [comment]
> > * Not sure what the current status of such a `/dev/null` symlink would mean, a misconfiguration probably. But certainly not `enabled`. `disabled` would probably be a better representation.
> 
> Indeed, considering the presence of a link as a sure fire sign a service is at all enabled, without looking at its contents, or where it points to, is dodgy at best.

If we consider such a `/dev/null` symlink in a .wants directory a misconfiguration, systemd should ignore it and log an error I think.

**Comment by twojstaryzdomu** [comment]
> > > * Not sure what the current status of such a `/dev/null` symlink would mean, a misconfiguration probably. But certainly not `enabled`. `disabled` would probably be a better representation.
> > 
> > 
> > Indeed, considering the presence of a link as a sure fire sign a service is at all enabled, without looking at its contents, or where it points to, is dodgy at best.
> 
> If we consider such a `/dev/null` symlink in a .wants directory a misconfiguration, systemd should ignore it and log an error I think.

Why? It is an empty file after all. It may be a result of informed user activity. It should be no different from the case when an empty file (not a link) is encountered.

To consider as service fully enabled, systemd needs to do due dilligence, tantamount to the find statements:
````
# touch a
# ln -s a b
# find -H a -samefile b
a
find -H b -samefile a
b
```
Without such due dilligence checks, a broken link (not necessarily one created with a informed purpose like in my use case) will falsely trigger a service enabled flag.

**Comment by twojstaryzdomu** [comment]
> > What if you need to use the service, just not under that specific `target/.wants` directory?
> 
> i cannot parse this. What is this supposed to mean?
> 
> You want to cancel individual deps? We have no mechanism for that. And its hard to do, since deps can be created in both directions and via various means.

Configuration in systemd is built up from fragments. Now, if fragments is permitted, there has to be some hierarchy, or an order,in which these get parsed. There has to be some handling for symlinks beyond e.g. a symlink for service X exists and we don't look at what it's pointing at at all.

For this issue specifically to be resolved, the premise is systemd needs to resolve the symlink and ensure it isn't a bad link `/dev/null`. An empty placeholder that has been put into place doesn't appear to be enabling a service. See below:

```
# cat > /lib/systemd/system/test.service << EOL
[Service]
ExecStart=true
[Install]
WantedBy=multi-user.target
EOL
# systemctl daemon-reload
# systemctl status test
○ test.service
     Loaded: loaded (/lib/systemd/system/test.service; disabled; vendor preset: enabled)
     Active: inactive (dead)
# ln -sf /dev/null /etc/systemd/system/multi-user.target.wants/test.service
# systemctl daemon-reload
# systemctl status test
○ test.service
     Loaded: loaded (/lib/systemd/system/test.service; enabled; vendor preset: enabled)
     Active: inactive (dead)
```
An empty file in place of a dependency link doesn't enable a service:
```
# rm /etc/systemd/system/multi-user.target.wants/test.service
# touch /etc/systemd/system/multi-user.target.wants/test.service
# systemctl daemon-reload
# systemctl status test
○ test.service
     Loaded: loaded (/lib/systemd/system/test.service; disabled; vendor preset: enabled)
     Active: inactive (dead)
```
A bad service file in place of a dependency link doesn't enabled a service:
```
# rm /etc/systemd/system/multi-user.target.wants/test.service
# echo test > /etc/systemd/system/multi-user.target.wants/test.service
# cat /etc/systemd/system/multi-user.target.wants/test.service
test
# systemctl daemon-reload
# systemctl status test
○ test.service
     Loaded: loaded (/lib/systemd/system/test.service; disabled; vendor preset: enabled)
     Active: inactive (dead)
```
Linking against a disabled service is triggering enabled status:
```
# rm /etc/systemd/system/multi-user.target.wants/test.service
# systemctl status ssh
○ ssh.service - OpenBSD Secure Shell server
     Loaded: loaded (/lib/systemd/system/ssh.service; disabled; vendor preset: enabled)
     Active: inactive (dead)
       Docs: man:sshd(8)
             man:sshd_config(5)
# ln -s /lib/systemd/system/ssh.service /etc/systemd/system/multi-user.target.wants/test.service
# systemctl daemon-reload
# systemctl status test
○ test.service
     Loaded: loaded (/lib/systemd/system/test.service; enabled; vendor preset: enabled)
     Active: inactive (dead)
```
A non-empty symlink pointing at a file is triggering enabled status:
```
# rm /etc/systemd/system/multi-user.target.wants/test.service
# ln -s /etc/passwd /etc/systemd/system/multi-user.target.wants/test.service
# systemctl daemon-reload
# systemctl status test
○ test.service
     Loaded: loaded (/lib/systemd/system/test.service; enabled; vendor preset: enabled)
     Active: inactive (dead)
```
Heck, even a symlink pointing at a directory is triggering enabled status:
```
# rm /etc/systemd/system/multi-user.target.wants/test.service
# ln -s /tmp /etc/systemd/system/multi-user.target.wants/test.service
# systemctl daemon-reload
# systemctl status test
○ test.service
     Loaded: loaded (/lib/systemd/system/test.service; enabled; vendor preset: enabled)
     Active: inactive (dead)
```
The systemd version is 251.2.

All these lead me to conclude symlinks aren't being examined, leading to possible issues when bad links (or merely pointing at an empty file i.e. `/dev/null`) are encountered.

As for #4830 being a duplicate, this issue is tangential, limited to symlinks specifically and as such smaller in scope. I'd be reluctant to close it. Including the problem under #4830 would amount to sweeping it under the carpet as #4830 has been outstanding for almost 7 years.

**Comment by poettering** [comment]
> Why error? It is an empty file after all. It may be a result of informed user activity. It should be no different from the case when an empty file (not a link) is encountered. Logging may be an option but there may be cases when it's intended (as in the case of the read-only `/usr/systemd`).

We do not load units through those symlinks, but we always load them from the main unit dirs, never the .wants/ subdirs. Or in other words, we use only the basename of the path stored in the symlinks themselves, and never open a unit file through the symlink. It's the only thing that makes sense really, because units typically have many deps that reference them, and if we loaded them through those symlinks then it would be undeterministic which definition we'd end up with...

**Comment by poettering** [comment]
> * The misleading status output. systemctl shows the service as enabled when actually it isn't
>       Not sure what the current status of such a `/dev/null` symlink would mean, a misconfiguration probably. But certainly not `enabled`. `disabled` would probably be a better representation.

The man pages already say that symlinks in .wants/ have to point to validly named unit files. /dev/null does not qualify. It's unspecified what it means if they point to random nonesense, but I am not sure this is a problem, if you do stuff that is clearly against what is documented.

That said, I#d be finde with just ignoring symlinks entirely that point to invalid stuff when determining whether a unit is enabled or no.

**Comment by poettering** [comment]
> 2\. Purposefully serving as service dependency black out (e.g. read-only `/lib/systemd` case) by pointing at an empty file (or `/dev/null`).

As mentioned this has been requested before, we do not support this right now. I'd be OK with adding a masking concept for deps, but it's not that easy to implement as you might think, because such masking should probably also override deps synthesized other ways then just .wants/ symlinks. Anyway, for that discussion → #4830

> I'd propose to forget about the symlink filename altogether and start looking at the symlink target to begin with. 

We look at symlink name and target. As I said.

**Comment by twojstaryzdomu** [comment]
> We use that because it allows you to instantiate stuff, ie. have a symlink from [foo@quux.service](mailto:foo@quux.service) to [foo@.service](mailto:foo@.service).

The above notwithstanding, systemd ought to ensure foo@instance.service is pointing at foo@.service, or report an error (misconfigured? blacked out?) and give up including that dependency.
